src/levanter/data/text.py

```diff
@@ -722,28 +796,69 @@ def _prepare_supervised_example(ex: dict, tokenizer: PreTrainedTokenizerBase, Po
         return lm_ex
 
 
-def mk_supervised_dataset(config: LMSupervisedDatasetConfig, tokenizer: PreTrainedTokenizerBase, Pos: hax.Axis):
-    import levanter.data
+def mk_supervised_datasets(
+    sources: Mapping[str, SupervisedSourceConfigBase] | SupervisedSourceConfigBase,
+    split: str,
+    tokenizer: PreTrainedTokenizerBase,
+    Pos: hax.Axis,
+) -> dict[str, tuple[AsyncDataset[LmExample], Sequence[str]]]:
+    """
+    Create supervised datasets from a set of sources.
+
+    Returns:
+        A dictionary of dataset names to tuples of the dataset and the tags associated with the dataset.
+    """
+    out: dict[str, tuple[AsyncDataset[LmExample], Sequence[str]]] = {}
+
+    if tokenizer.pad_token is None:
+        tokenizer.pad_token = tokenizer.eos_token
+
+    if isinstance(sources, Mapping):
+        for name, config in sources.items():
+            source = config.get_shard_source(split)
+            if source is None:
+                continue
+
+            ds = _cache_supervised_set(
+                source, config.cache_dir, tokenizer, Pos, config.input_field, config.output_field
+            )
+
+            if config.tags is None:
+                tags = [name]
+            else:
+                tags = config.tags + [name]
 
-    # Choose data source based on config
-    if config.hf_dataset_name is not None:
-        # Using HF dataset
-        dataset = levanter.data.datasource_from_hf(config.hf_dataset_name, split=config.hf_dataset_split)
+            out[name] = (ds, tags)
     else:
-        # Using local files
-        validation_urls = [url for url_pat in config.validation_urls for url in fsspec_utils.expand_glob(url_pat)]
-        if not validation_urls:
-            raise ValueError("Must specify either hf_dataset_name or validation_urls")
-        dataset = levanter.data.datasource_from_jsonl(validation_urls)
+        source = sources.get_shard_source(split)  # type: ignore
+        if source is not None:
+            ds = _cache_supervised_set(
+                source, sources.cache_dir, tokenizer, Pos, sources.input_field, sources.output_field
+            )
+            tags = sources.tags or []
+            if isinstance(sources, SupervisedHfSourceConfig):
+                name = sources.id
+                if sources.name is not None:
+                    name = f"{name}/{sources.name}"
+
+                tags = [name] + tags
+            else:
+                name = "default"
+            out[name] = (ds, tags)
+
+    return out
+
+
+def mk_supervised_dataset(
```

Comment:
Seems like we are replacing this with mk_supervised_datasets? Do we still need this function then?

Reply:
it's still used in sft.py atm but i'd like to remove it

Reply:
Sounds good